src/wic/plugins.py

```diff
@@ -117,6 +127,51 @@ def get_tools_cwl(homedir: str, validate_plugins: bool = False,
     return tools_cwl
 
 
+def cwl_update_outputs_optional(cwl: Cwl) -> Cwl:
+    """Updates outputs as optional (if any)
+
+    Args:
+        cwl (Cwl): A CWL CommandLineTool
+
+    Returns:
+        Cwl: A CWL CommandLineTool with outputs optional (if any)
+    """
+    cwl_mod = copy.deepcopy(cwl)
+    # Update version
+    # when clauses require cwl v1.2
+    # See https://www.commonwl.org/v1.2/Workflow.html#Conditional_execution_(Optional)
+    cwl_mod['cwlVersion'] = 'v1.2'
+    # Update success codes to allow simple failures
+    cwl_mod['successCodes'] = [0, 1]
+    # Update outputs optional
+    for out_key, out_val_dict in cwl_mod['outputs'].items():
+        if isinstance(out_val_dict['type'], str) and out_val_dict['type'][-1] != '?':
+            out_val_dict['type'] += '?'
+    return cwl_mod
+
+
+def cwl_update_outputs_optional_rosetree(rose_tree: RoseTree) -> RoseTree:
+    """Updates outputs optional for every CWL CommandLineTool
+
+    Args:
+        rose_tree (RoseTree): The RoseTree returned from compile_workflow(...).rose_tree
+
+    Returns:
+        RoseTree: rose_tree with output optional updates to every CWL CommandLineTool
+    """
+    n_d: NodeData = rose_tree.data
+    # NOTE: Since only class: CommandLineTool should have dockerPull tags,
```

Comment:
This comment has been copy & pasted without considering whether the statement still applies for the cwl_update_outputs_optional function.

Reply:
Specifically, test_rand_fail.yml does not contain any subworkflows, so this code path is not exercised. I recommend extracting the fail step into a single-step subworkflow fail.yml